Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,167 @@
+        else
+            marginBoxSize = child->logicalHeight() + child->marginLogicalHeight();
+
+    } else {
+        if (GridLayoutFunctions::isOrthogonalChild(m_renderGrid, *child))
+            marginBoxSize = child->isHorizontalWritingMode() ? child->height() + child->verticalMarginExtent() : child->width() + child->horizontalMarginExtent();
+        else
+            marginBoxSize = child->logicalWidth() + child->marginLogicalWidth();
+    }
+    return marginBoxSize;
+}
+void GridMasonryLayout::updateRunningPositions(const RenderBox* child, const GridArea& area)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? area.columns : area.rows;
+    ASSERT(gridSpan.startLine() < m_runningPositions.size() && gridSpan.endLine() <= m_runningPositions.size());
+    gridSpan.clamp(m_runningPositions.size());
+
+    LayoutUnit previousRunningPosition;
+    for (auto line : gridSpan)
+        previousRunningPosition = std::max(previousRunningPosition, m_runningPositions[line]);
+
+    auto newRunningPosition = masonryAxisMarginBoxForItem(child) + previousRunningPosition + m_masonryAxisGridGap; 
+    m_gridContentSize = std::max(m_gridContentSize, newRunningPosition - m_masonryAxisGridGap);
+
+    for (auto span : gridSpan)
+        m_runningPositions[span] = std::max(m_runningPositions[span], newRunningPosition);
+
+    updateItemOffset(child, previousRunningPosition);
+}
+
+void GridMasonryLayout::updateItemOffset(const RenderBox* child, LayoutUnit offset)
+{
+    m_itemOffsets.add(child, offset);
+}
+
+GridArea GridMasonryLayout::nextMasonryPositionForItem(const RenderBox* item)
+{
+    auto itemSpanLength = GridPositionsResolver::spanSizeForAutoPlacedItem(*item, m_gridAxisDirection);
+    auto gridAxisLines = m_gridAxisTracksCount + 1;
+
+    LayoutUnit smallestMaxPos = LayoutUnit::max();
+    unsigned smallestMaxPosLine = 0;
+    for (unsigned startingLine = 0; startingLine < gridAxisLines - itemSpanLength; startingLine++) {
+        LayoutUnit maxPosForCurrentStartingLine;
+        for (unsigned lineOffset = 0; lineOffset < itemSpanLength; lineOffset++)
+            maxPosForCurrentStartingLine = std::max(maxPosForCurrentStartingLine, m_runningPositions[startingLine + lineOffset]);
+        if (maxPosForCurrentStartingLine < smallestMaxPos) {
+            smallestMaxPos = maxPosForCurrentStartingLine;
+            smallestMaxPosLine = startingLine;
+        } 
+    }
+
+    GridSpan masonryAxisSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+    GridSpan gridAxisSpan = GridSpan::translatedDefiniteGridSpan(smallestMaxPosLine, smallestMaxPosLine + itemSpanLength);
+    return m_masonryAxisDirection == ForRows ? GridArea { masonryAxisSpan, gridAxisSpan } : GridArea { gridAxisSpan, masonryAxisSpan };
+}
+
+LayoutUnit GridMasonryLayout::offsetForChild(const RenderBox& child) const
+{
+    if (m_itemOffsets.contains(&child))
+        return m_itemOffsets.get(&child);
```

Comment:
Nit: This pattern causes two lookups in the hash table (one for 'contains', another for 'get'). If you want to use the result of the lookup, it's better to use 'find':

```
const auto& offsetIter = m_itemOffsets.find(&child);
if (offsetIter == m_itemOffsets.end())
    return 0_lu;
return *offsetIter;
```

There might be a` findOr<>` thing that would work, too.